### PR TITLE
Add a first-cut implementation of /interop

### PR DESCRIPTION
**Pull request by lukebjerring**
This is essentially minor tweaks to get https://github.com/mdittmer/wptdashboard/pull/3 sync'd with the current state of WPTD.

Running @ https://lukebjerring-interop-dot-wptdashboard.appspot.com/interop/

### CODE REVIEW

**Comment by rwaldron** [comment]
Related to this: https://github.com/w3c/wptdashboard/issues/374

**Comment by rwaldron** [comment]
Also, just a heads up that I'm working on reviewing this now

File: webapp/components/interop.html [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+    }
+    const match = search.match(/[^a-zA-Z0-9%]q=([^&]+)/);
+    if (!match) {
+      return '';
+    }
+    const subMatch = match[1];
+    const decodedMatch = window.decodeURIComponent(subMatch);
+    return decodedMatch;
+  };
+
```
Comment (rwaldron) [inline webapp/components/interop.html]: All of the `_` prefixed methods need to be renamed without a `_` prefix. Once the testing work lands, this is going to be enforced via CI linting.

If some method or property needs to be protected from accidental access, use Symbol

File: webapp/components/interop.html [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+        </tr>
+      </template>
+    </tbody>
+  </table>
+
+</template>
+
+<script>
+  'use strict';
+  const urlToMetricsPath = url => url.pathname.replace(/\/interop(.+)(\/)?$/, '$1');
```
Comment (rwaldron) [inline webapp/components/interop.html]: `url` => `location`

File: webapp/components/interop.html [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+  const urlToQuery = url => {
```
Comment (rwaldron) [inline webapp/components/interop.html]: `url` => `location`

File: webapp/components/interop.html [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+    // Drop initial "?" from url.search.
+    const search = url.search;
+    if (!search) {
+      return '';
+    }
+    const match = search.match(/[^a-zA-Z0-9%]q=([^&]+)/);
```
Comment (rwaldron) [inline webapp/components/interop.html]: I would like to see a comprehensive data set containing example urls that are expected to match.

File: webapp/components/interop.html [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    if (!match) {
+      return '';
+    }
+    const subMatch = match[1];
```
Comment (rwaldron) [inline webapp/components/interop.html]: This assignment serves no purpose, `match[1]` could be passed directly, ie. ` window.decodeURIComponent(match[1]);`

File: webapp/components/interop.html [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+    const decodedMatch = window.decodeURIComponent(subMatch);
```
Comment (rwaldron) [inline webapp/components/interop.html]: This assignment also serves no purpose, just `return window.decodeURIComponent(match[1]);`

File: webapp/components/interop.html [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+      }
+      const encodedQuery = window.encodeURIComponent(query);
+      stateFunc(state, '', `/interop${path}?q=${encodedQuery}`);
+    }
+
+    _computePathIsATestFile(path) {
+      return path.endsWith('.html') || path.endsWith('.htm') ||
+              path.endsWith('.py') || path.endsWith('.svg') ||
+              path.endsWith('.xhtml') || path.endsWith('.xht') ||
+              path.endsWith('.xml');
```
Comment (rwaldron) [inline webapp/components/interop.html]: `return /(\.(html|htm|py|svg|xhtml|xht|xml)$)/.test(path);`

File: webapp/components/interop.html (hunk 4 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: @mdittmer

File: webapp/components/interop.html (hunk 7 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+    }
+
+    _computeThLabels (allMetrics) {
+      const numLabels = allMetrics.metadata.test_runs.length + 1;
+      let labels = new Array();
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+      } catch (e) {
+        test = str => str.indexOf(query) !== -1
+      }
+      return this._shouldQueryAll()
+        ? test(subdir)
+        : subdir.indexOf('/') === -1 && test(subdir);
+    }
+
+    _relPath(path) {
+      return path.replace(this.path + '/', '');
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+        _queryTimeoutId: {
+          type: Number,
+          value: 0
+        },
+      }
+    }
+
+    async connectedCallback() {
+      await super.connectedCallback();
+      window.addEventListener('popstate', event => {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html (hunk 1 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html (hunk 5 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Fun fact: Intellij has a neat 'inline' refactor method under ⌥⌘N

File: webapp/components/interop.html (hunk 6 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html (hunk 3 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html (hunk 2 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

**Comment by lukebjerring** [comment]
Changes added + https://lukebjerring-interop-dot-wptdashboard.appspot.com/interop/ redeployed. PTAL.

**Comment by boazsender** [comment]
I'm worried about the route for this. The compat spec could have just as easily been called the interop spec, and we have tests for it: https://wpt.fyi/compat.

**Comment by rwaldron** [comment]
@boazsender yes, as I mentioned here: https://github.com/w3c/wptdashboard/issues/374 (you should add the case you found to _that_ issue)

**Comment by rwaldron** [comment]
Presently there are leading forward slashes on the dir name links: 



But we changed that to trailing slashes in the main view: 



But also they are totally missing from nested dir view display:

File: webapp/components/interop.html [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+      <input value="{{ query::input }}"
+             class="query"
+             placeholder="{{ queryPlaceholder }}"
+             data-foo="bar">
+    </template>
+    <div class="path">
+      <a href="/interop/" on-click="_navigate">WPT</a>
+      <template is="dom-repeat" items="{{ splitIntoLinkedParts(path) }}" as="part">
+        <span class="path-separator">/</span>
+        <a href="/interop{{ part.path }}"  on-click="_navigate">{{ part.name }}</a>
```
Comment (rwaldron) [inline webapp/components/interop.html]: `_navigate` => `navigate`

File: webapp/components/interop.html [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+  <template is="dom-if" if="{{ pathIsATestFile }}">
+    <div class="links">
+      <ul>
+        <li>
+          <a href$="https://github.com/w3c/web-platform-tests/blob/master[[path]]"
+             target="_blank">View source on GitHub</a></li>
+        <li><a href$="https://w3c-test.org[[path]]" target="_blank">Run in your
+          browser on w3c-test.org</a></li>
+      </ul>
+    </div>
```
Comment (rwaldron) [inline webapp/components/interop.html]: This is nearly identical to: https://github.com/w3c/wptdashboard/blob/d7354d5a4c3735121cb37c845217039a3bdcb1f5/webapp/components/wpt-results.html#L117-L122

File: webapp/components/interop.html [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+    computeThLabels (allMetrics) {
+      const numLabels = allMetrics.metadata.test_runs.length + 1;
+      let labels = [];
+      for (let i = 0; i < numLabels; i++) {
+        labels[i] = `${i} / ${numLabels - 1} browsers`;
+      }
+      return labels;
+    }
+
+    computeMetrics (path, pathIsATestFile, query, allMetrics, currentNode) {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/interop.html]: Cleaned up pathIsATestFile & currentNode, but the 'this' object doesn't have those properties available in the scope of the compute function.

File: webapp/components/interop.html [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+    <tbody>
+      <template is="dom-repeat" items="{{ metrics.data }}" as="node">
+        <tr>
+          <td>
+            <template is="dom-if" if="{{ !computePathIsATestFile(node.dir) }}">
+              <b><a href="/interop{{ node.dir }}" on-click="navigate">{{ relPath(node.dir) }}</a></b>
+            </template>
+            <template is="dom-if" if="{{ computePathIsATestFile(node.dir) }}">
+              <a href="/interop{{ node.dir }}" on-click="navigate">{{ relPath(node.dir) }}</a>
+            </template>
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html (hunk 11 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Done.

File: webapp/components/interop.html (hunk 12 above)
Comment (lukebjerring) [reply webapp/components/interop.html]: Noted; Will split into a component in separate PR.

File: webapp/components/interop.html [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+      const rounded = Math.round(fraction * 1000) / 1000;
+      return `background-color: rgba(170, 136, 255, ${rounded})`;
+    }
+
+    queryChanged (_, oldQuery) {
+      // Debounce first initialization.
+      if (oldQuery === undefined) {
+        return
+      }
+      if (this._queryTimeoutId) {
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply webapp/components/interop.html]: Done (as best I can understand Symbol's intended use?)

**Comment by lukebjerring** [comment]
Changes pushed + redeployed (@ https://lukebjerring-interop-dot-wptdashboard.appspot.com/interop) - PTAL.

File: webapp/components/interop.html [hunk 16]
```diff
@@ -206,6 +203,8 @@ <h2>Interoperability</h2>
     return window.decodeURIComponent(match[1]);
   };
 
+  const queryTimeoutId = Symbol('queryTimeoutId');
```
Comment (rwaldron) [inline webapp/components/interop.html]: ❤️ ❤️ ❤️ ❤️ ❤️ ❤️

**Pull request merged**